tests/std/tests/P0896R4_views_join/test.cpp

```diff
@@ -0,0 +1,505 @@
+            }
+
+            if constexpr (copyable<V>) {
+                auto r2                              = r;
+                const same_as<iterator_t<R>> auto i2 = r2.begin();
+                if (!is_empty) {
+                    assert(*i2 == *i);
+                }
+            }
+
+            if constexpr (CanMemberBegin<const R> && !CanBegin<const R>) {
+                static_assert(input_iterator<decltype(declval<const R>().begin())>);
+                static_assert(sentinel_for<decltype(declval<const R>().end()), decltype(declval<const R>().begin())>);
+            }
+
+            if constexpr (CanBegin<const R> && forward_range<R>) {
+                const iterator_t<const R> ci = as_const(r).begin();
+                if (!is_empty) {
+                    assert(*ci == *i);
+                }
+
+                if constexpr (copyable<V>) {
+                    const auto r2                               = r;
+                    const same_as<iterator_t<const V>> auto ci2 = r2.begin();
+                    if (!is_empty) {
+                        assert(*ci2 == *i);
+                    }
+                }
+            }
+        }
+
+        // Validate join_view::end
+        static_assert(CanMemberEnd<R>);
+        static_assert(CanMemberEnd<const R> == (input_range<const V> && is_reference_v<range_reference_t<const V>>) );
+        // clang-format off
+        static_assert(common_range<R> == (forward_range<V> && is_reference_v<range_reference_t<V>> && common_range<V>
+                                          && forward_range<Inner> && common_range<Inner>) );
+        static_assert(common_range<const R> == (forward_range<const V> && is_reference_v<range_reference_t<const V>>
+                                                && common_range<const V> && forward_range<range_reference_t<const V>>
+                                                && common_range<range_reference_t<const V>>) );
+        // clang-format on
+        if (!is_empty) {
+            const ranges::sentinel_t<R> s = r.end();
+            if constexpr (bidirectional_range<R> && common_range<R>) {
+                assert(*prev(s) == *prev(end(expected)));
+
+                if constexpr (copyable<V>) {
+                    auto r2 = r;
+                    assert(*prev(r2.end()) == *prev(end(expected)));
+                }
+            }
+
+            if constexpr (CanMemberEnd<const R>) {
+                const ranges::sentinel_t<const R> cs = as_const(r).end();
+                if constexpr (bidirectional_range<R> && common_range<R>) {
+                    assert(*prev(cs) == *prev(end(expected)));
+
+                    if constexpr (copyable<V>) {
+                        const auto r2                   = r;
+                        const ranges::sentinel_t<R> cs2 = r2.end();
+                        assert(*prev(cs2) == *prev(end(expected)));
+                    }
+                }
+            }
+        }

```

Comment:
I have no Idea why I structured it differently be I think we should have the same structure as above.